File: leetcode/java/0121. Best Time to Buy and Sell Stock.java
package leetcode.java;

// Time complexity: O(N)
// Space complexity: O(1)

class Solution {
    public int maxProfit(int[] prices) {
        
        int lowest = prices[0];
        int maxProfit = 0;
        for (int price : prices) {
            if (price > lowest) {
                maxProfit = maxProfit > price - lowest ? maxProfit : price - lowest;
            } else {
                lowest = price;
            }
        }
        return maxProfit;
    }
}

File: leetcode/python/0121. Best Time to Buy and Sell Stock.py
# Time Complexity: O(N)
# Space complexity: O(1)

class Solution(object):
    def maxProfit(self, prices):
        minimum = 99999
        max_price = 0
        for p in prices:
            if p < minimum:
                minimum = p
            else:
                max_price = max(max_price, p - minimum)
        return max_price